fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java

```diff
@@ -0,0 +1,458 @@
+            LogicalAggregationOperator aggOp = optExpression.getOp().cast();
+            // distinct/count* aggregate can't push down
+            if (aggOp.getAggregations().values().stream().anyMatch(c -> c.isDistinct() || c.isCountStar())) {
+                return visit(optExpression, context);
+            }
+
+            // all constant can't push down
+            if (!aggOp.getAggregations().isEmpty() &&
+                    aggOp.getAggregations().values().stream().allMatch(ScalarOperator::isConstant)) {
+                return visit(optExpression, context);
+            }
+
+            // no-group-by don't push down
+            if (aggOp.getGroupingKeys().isEmpty()) {
+                return visit(optExpression, context);
+            }
+
+            context = new AggregatePushDownContext();
+            context.setAggregator(aggOp);
+            return context;
+        }
+
+        @Override
+        public AggregatePushDownContext visitLogicalWindow(OptExpression optExpression,
+                                                           AggregatePushDownContext context) {
+            if (canNotPushDown(optExpression, context)) {
+                return visit(optExpression, context);
+            }
+            LogicalWindowOperator windowOp = optExpression.getOp().cast();
+            // only support distinct aggregation push down window
+            if (!context.origAggregator.getAggregations().isEmpty() || !windowOp.canPushDownAgg()) {
+                return visit(optExpression, context);
+            }
+
+            List<ColumnRefOperator> additionalGroupBys = Lists.newArrayList();
+            windowOp.getPartitionExpressions().stream()
+                    .flatMap(e -> e.getUsedColumns().getStream().mapToObj(factory::getColumnRef)).forEach(
+                            additionalGroupBys::add);
+
+            windowOp.getWindowCall().keySet().forEach(context.groupBys::remove);
+            windowOp.getOrderByElements().forEach(o -> additionalGroupBys.add(o.getColumnRef()));
+            additionalGroupBys.forEach(c -> context.groupBys.put(c, c));
+            context.aggregations.putAll(windowOp.getWindowCall());
+            context.hasWindow = true;
+            return context;
+        }
+
+        @Override
+        public AggregatePushDownContext visitLogicalProject(OptExpression optExpression,
+                                                            AggregatePushDownContext context) {
+            if (canNotPushDown(optExpression, context)) {
+                return visit(optExpression, context);
+            }
+
+            LogicalProjectOperator projectOp = optExpression.getOp().cast();
+
+            if (projectOp.getColumnRefMap().entrySet().stream().allMatch(e -> e.getValue().equals(e.getKey()))) {
+                return context;
+            }
+
+            // rewrite
+            ReplaceColumnRefRewriter rewriter = new ReplaceColumnRefRewriter(projectOp.getColumnRefMap());
```

Comment:
It seems there is special process for `case when` and `if` fucntion in `PushDownAggregateRewriter.rewriteProject()`. Should we consider it here?

Reply:
case when and if function will be processed in the future, but it is quite different from PushDownAggregateRewriter's present way.  A right way is to allow push-down agg to sits above union/project/join operator.